cupyx/scipy/sparse/construct.py

```diff
@@ -88,6 +88,234 @@ def spdiags(data, diags, m, n, format=None):
+
+    blocks_flat = []
+    for m in range(M):
+        for n in range(N):
+            if blocks[m][n] is not None:
+                blocks_flat.append(blocks[m][n])
+
+    if len(blocks_flat) == 0:
+        return coo.coo_matrix((0, 0), dtype=dtype)
+
+    # if blocks.ndim != 2:
+    #     raise ValueError('blocks must be 2-D')
+
+    # check for fast path cases
+    if (N == 1 and format in (None, 'csr') and
+            all(isinstance(b, csr.csr_matrix)
+                for b in blocks_flat)):
+        A = _compressed_sparse_stack(blocks_flat, 0)
+        if dtype is not None:
+            A = A.astype(dtype)
+        return A
+    elif (M == 1 and format in (None, 'csc')
+          and all(isinstance(b, csc.csc_matrix) for b in blocks_flat)):
+        A = _compressed_sparse_stack(blocks_flat, 1)
+        if dtype is not None:
+            A = A.astype(dtype)
+        return A
+
+    block_mask = cupy.zeros((M, N), dtype=bool)
+    brow_lengths = cupy.zeros(M+1, dtype=cupy.int64)
+    bcol_lengths = cupy.zeros(N+1, dtype=cupy.int64)
+
+    # convert everything to COO format
+    for i in range(M):
+        for j in range(N):
+            if blocks[i][j] is not None:
+                A = coo.coo_matrix(blocks[i][j])
+                blocks[i][j] = A
+                block_mask[i][j] = True
+
+                if brow_lengths[i+1] == 0:
+                    brow_lengths[i+1] = A.shape[0]
+                elif brow_lengths[i+1] != A.shape[0]:
+                    msg = ('blocks[{i},:] has incompatible row dimensions. '
+                           'Got blocks[{i},{j}].shape[0] == {got}, '
+                           'expected {exp}.'.format(i=i, j=j,
+                                                    exp=brow_lengths[i+1],
+                                                    got=A.shape[0]))
+                    raise ValueError(msg)
+
+                if bcol_lengths[j+1] == 0:
+                    bcol_lengths[j+1] = A.shape[1]
+                elif bcol_lengths[j+1] != A.shape[1]:
+                    msg = ('blocks[:,{j}] has incompatible row dimensions. '
+                           'Got blocks[{i},{j}].shape[1] == {got}, '
+                           'expected {exp}.'.format(i=i, j=j,
+                                                    exp=bcol_lengths[j+1],
+                                                    got=A.shape[1]))
+                    raise ValueError(msg)
+
+    nnz = sum(block.nnz for block in blocks_flat)
+    if dtype is None:
+        all_dtypes = [blk.dtype for blk in blocks_flat]
+        dtype = upcast(*all_dtypes) if all_dtypes else None
+
+    row_offsets = cupy.cumsum(brow_lengths)
+    col_offsets = cupy.cumsum(bcol_lengths)
+
+    shape = (row_offsets[-1], col_offsets[-1])
+
+    data = cupy.empty(nnz, dtype=dtype)
+    idx_dtype = get_index_dtype(maxval=max(shape))
+    row = cupy.empty(nnz, dtype=idx_dtype)
+    col = cupy.empty(nnz, dtype=idx_dtype)
+
+    nnz = 0
+    ii, jj = cupy.nonzero(block_mask)
```

Comment:
`cupy.nonzero` synchronizes the device, which significantly affects performance (it's not documented yet: issue #2797). Is it possible to avoid using it?

Reply:
I have not dived extremely deep into the implementation of `nonzero` yet, but I see a reverse slicing of the array with a comment next to it that it's being synchronized. This raises two questions:

1. Is the `int()` implicitly causing a copy to host, which is causing the need for the synchronization?
2. A full device synchronization is very expensive here. If it's absolutely necessary, is it possible just to do a stream synchronize here?

Reply:
>1. Is the `int()` implicitly causing a copy to host, which is causing the need for the synchronization?

Yes, that's correct.

>2. A full device synchronization is very expensive here. If it's absolutely necessary, is it possible just to do a stream synchronize here?

`int()` (or `nonzero`) only synchronizes the current stream (if it is set).

If it's difficult to remove synchronization, it's OK to leave it, but could you add a comment that it synchronizes?